src/QMCHamiltonians/MomentumEstimator.cpp

```diff
@@ -172,115 +147,271 @@ void MomentumEstimator::setParticlePropertyList(PropertySetType& plist
   if (!hdf5_out)
   {
     copy(nofK.begin(),nofK.end(),plist.begin()+myIndex+offset);
-    copy(compQ.begin(),compQ.end(),plist.begin()+myIndex+nofK.size()+offset);
   }
 }
 
 bool MomentumEstimator::putSpecial(xmlNodePtr cur, ParticleSet& elns, bool rootNode)
 {
   OhmmsAttributeSet pAttrib;
   std::string hdf5_flag="yes";
+  kmax=0.0;
+  kmax0=0.0;
+  kmax1=0.0;
+  kmax2=0.0;
   pAttrib.add(hdf5_flag,"hdf5");
-  pAttrib.add(kgrid,"grid");
-  pAttrib.add(M,"samples");
+  //pAttrib.add(kgrid,"grid");
+  pAttrib.add(kmax,"kmax");
+  pAttrib.add(kmax0,"kmax0");
+  pAttrib.add(kmax1,"kmax1");
+  pAttrib.add(kmax2,"kmax2");
+  pAttrib.add(M,"samples"); // default value is 40 (in the constructor)
   pAttrib.put(cur);
   hdf5_out = (hdf5_flag=="yes");
+  RealType min_Length=1.0e10;
+  PosType vec_length(0);
+  for (int i=0; i<OHMMS_DIM; i++)
+  {
+    PosType a(0);
+    for (int j=0; j<OHMMS_DIM; j++)
+      a[j]=elns.Lattice.G(j,i);
+    //length of reciprocal lattice vector
+    vec_length[i]=2.0*M_PI*std::sqrt(dot(a,a));
+    if (min_Length>vec_length[i])
+      min_Length=vec_length[i];
+  }
```

Comment:
minimal G vector length may not be a good choice. I feel more safe to use WS radius in G space.

Reply:
This is only used for getting an overestimation for the for-loop bounds such that definitely all desired k-points will be collected. I suspect that WS radius would not be adequate here.

Reply:
Shortest length k-space axis => largest needed grid size

Reply:
Example,
```
g0=[0.5  0.0  0.0]
g1=[0.5  0.3  0.0]
g2=[0.0  0.0  1.0]
```
Shortest k-space axis is 0.5 but the WS_radius=0.3, the shortest vector is actually g1-g0.

Reply:
If it always leads to a smaller value, then it should be ok to use that also, but not necessary. Since kmax0 refers to the maximum k-value along the g0 direction, then the nmax0 would be less than int(kmax0/|g0|)+1. The same applies to all directions. Now, if one chooses the maximum of those nmax0, nmax1, and nmax2, then one should definitely go through all possible k-points... also the shortest vector you showed will be included, since the loop goes from -nmax to nmax.

Reply:
In principle, I don't think -nmax to nmax is sufficient if nmax is defined by the min_length.
```
g0=[0.5  0.0  0.0]
g1=[0.5  0.1  0.0]
g2=[0.0  0.0  1.0]
```
g1-g0=0.1 and you need 5x nmax to include all the points in the sphere.
However, in QMCPACK we use translation symmetry only up to 27 corners. So probably if such extreme case happens, many parts of the code already break down. So probably using the min_Length is OK within our assumption.

Reply:
Yes, you are right. In general it will be better to use the WS_radius.